OEP-6/OEP-6.mediawiki

```diff
@@ -0,0 +1,394 @@
+
+<pre>
+interface Caller {
+  url?: string;
+  id?: string;
+}
+</pre>
+
+===Encoding===
+The interaction with dAPI often requires specification of '''Address''', '''Identity''', '''Contract address''' or '''Public Key'''. Those concepts are represented by string representation with specific encodings.
+
+====Address====
+* 34 bytes long
+* using Base58 check encoding
+* starts with A letter
+
+====Contract address====
+* 40 bytes long
+* using Hex encoding
+
+====Identity====
+* 42 bytes long
+* using Base58 check encoding for 34 byte suffix
+* starts with 'did:ont:' prefix
+
+====Public Key====
+* 33/35 bytes long
+* using Hex encoding
+
+===Complex structures===
+All the functions of this dAPI, which requires arguments are using one typed argument instead of many. The reasons are better readability and future extensibility. 
+
+API specification is a complex document. Every method has some inputs and outputs. Although we use the primitive types (numbers, strings, booleans) whenever possible, there are places where a complex object is required. To precisely describe the structure of those objects, we'd chosen Typescript-like syntax.
+
+===Components===
+Although this proposal is bringing clear and simple API for the dApps, the individual functions can be divided into these components:
+
+* '''Network''', a thin wrapper around the Ontology Node API, masking the complexity of rpc/rest calls and web-sockets with Request-Response facade.
+
+* '''Provider''', functions for getting information about dAPI provider
+
+* '''Asset''', functions for transferring assets between user account and others.
+
+* '''Identity''', functions for interacting with own ONT-ID identity.
+
+* '''SmartContract''', a high level wrapper around the Smart Contract invocation and deployment.
+
+* '''Message''', functions for signing arbitrary messages.
+
+* '''Utils''', a group of utility function for encoding and decoding the data from/to blockchain.
+
+===Network===
+A network API consists of:
+
+<pre>
+type NetworkType = 'MAIN' | 'TEST' | 'PRIVATE';
+type Asset = 'ONT' | 'ONG';
+
+interface Network {
+  type: NetworkType;
+  address: string;
+}
+
+function getGenerateBlockTime(): Promise<number | null>
+function getNodeCount(): Promise<number>
+function getBlockHeight(): Promise<number>
+function getMerkleProof({ txHash: string }): Promise<MerkleProof>
+function getStorage({ contract: string, key: string }): Promise<string>
+function getAllowance({ asset: Asset, fromAddress: string, toAddress: string }): Promise<number>
+function getBlock({ block: number | string }): Promise<Block>
+function getTransaction({ txHash: string }): Promise<Transaction>
+function getNetwork(): Network
```

Comment:
Should there also be a method for requesting that the wallet change networks? Lets say that the dApp has a setting to run on testnet. Would be good to allow the dApp to request that the provider use the testnet, as it would be terrible if transactions that were meant to be sent to testnet by the dApp, were instead sent to mainnet by the provider.

Reply:
I don't see much point in requesting dAPI provider to change network. dApp has information about used network. It could warn (or even forbid the action) the user that he is calling the sc/transfer on MainNet or instruct the user to change the network.

Reply:
Then how should dApps coordinate using the test net over main net for their transactions? Are we assuming that this protocol will always be used with main net? Will users have to make sure themselves that their provider wallet and the dApp are both set to test net or both set to main net?

Reply:
I don't think a dApp should communicate something on MainNet and something on TestNet for single user account. A single dApp should communicate on one network only. 

But it does not mean, it can work on both Nets. It just means, that the user account on TestNet and MainNet are two different accounts. 

About coordinating the dApp and Provider net. I think the Provider is the one here, who will decide on which network are the dApp currently running. dApp should detect the Net from Provider, and act accordingly (there are more options):

1. show information about the network and inform the user about running on specific Net (e.g: a yellow bar informing the user, that he is on testNet)

2. don't allow user to do anything and inform him about necessity of switching network, because the dApp SC is deployed only to specific Net.

3. don't care at all, and only distinguishing the user accounts between Nets.